Adds mc batch commands to the docs (#620)
minio/docs: source/administration/batch-framework.rst

.. _minio-batch-framework:

===============
Batch Framework
===============

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 2

.. versionchanged:: MinIO RELEASE.2022-10-08T20-11-00Z or later

   The Batch Framework was introduced with the ``replicate`` job type in the :mc:`mc` RELEASES.2022-10-08T20-11-00Z.

Overview
--------

> The MinIO Batch Framework allows you to create, manage, monitor, and execute jobs using a YAML file description.

Reviewer: YAML-formatted job definition ?

Author: Fixed.

> Running on the MinIO deployment takes advantage of the cluster's server-side processing power without constraints of the local machine where you run the :ref:`MinIO Client <minio-client>`.

Reviewer: Running batch jobs?

Author: Why, yes. Yes, that's a good idea.

> A batch job definition file may contain definitions for a single job task or many job tasks.

Reviewer: bit repetitive. Spitballing:

```suggestion
A batch definition can contain any number of job tasks.
MinIO does not impose any limit on the number of job tasks per batch definition
```

I'm still chewing on our terminology here:

- Batch Job, which consists of a
- Definition File, which consists of one or more
- Jobs

Something here feels a little inefficient, but I can't think of a better way to do it.

Author: It's a `Batch file` which is a YAML-formatted definition of one or more `Jobs`.

Author: Gonna try to clean this up throughout.

MinIO does not have a pre-defined limit on the number of job tasks that you can define in a definition file.

> Initially, the framework allows for creating ``replicate`` jobs to manage bucket replication tasks.

Reviewer: We can simplify as:

"MinIO supports the following batch job types:

- `replicate` - <descriptor>

That way it's easy when we add more. Drop 'MinIO anticipates' because maybe we never will :)

Author: Created as a table.
Easy to add more rows later, or not.

MinIO anticipates adding additional job types to the framework in future releases.

> The MinIO Batch Framework requires that you have :ref:`installed the MinIO Client <mc-install>` and defined the :mc-cmd:`alias <mc alias set>` for the MinIO deployment(s) where you will run the job definition.

Reviewer: I think the intent here is that at least for now you can only create and manage batch jobs using the MinIO Client.

It's probably enough to just state "You can create and manage batch jobs using the :mc:`mc batch` command"

Author: Created a Prereq section instead with two bullet points

The :mc:`mc batch` commands include

.. list-table::
   :widths: 30 70
   :width: 90%

   * - :mc:`mc batch generate`
     - .. include:: /reference/minio-mc/mc-batch-generate.rst
          :start-after: start-mc-batch-generate-desc
          :end-before: end-mc-batch-generate-desc
   * - :mc:`mc batch start`
     - .. include:: /reference/minio-mc/mc-batch-start.rst
          :start-after: start-mc-batch-start-desc
          :end-before: end-mc-batch-start-desc
   * - :mc:`mc batch list`
     - .. include:: /reference/minio-mc/mc-batch-list.rst
          :start-after: start-mc-batch-list-desc
          :end-before: end-mc-batch-list-desc
   * - :mc:`mc batch status`
     - .. include:: /reference/minio-mc/mc-batch-status.rst
          :start-after: start-mc-batch-status-desc
          :end-before: end-mc-batch-status-desc
   * - :mc:`mc batch describe`
     - .. include:: /reference/minio-mc/mc-batch-describe.rst
          :start-after: start-mc-batch-describe-desc
          :end-before: end-mc-batch-describe-desc

Job Types
---------

> Currently, the MinIO Batch Framework supports the following type of job:

Reviewer: ```suggestion
The MinIO Batch Framework supports the following job types:
```

Author: Done.

Author: Actually, this was repetitive. Deleted since we already said this above.

- ``replicate``

``replicate``
~~~~~~~~~~~~~

The ``replicate`` job type performs a single-run replication of objects from one MinIO deployment or location to another MinIO deployment or location.
The end result of the job is comparable to running :ref:`bucket replication <minio-bucket-replication>`. 

> Key advantages of doing replication with a batch job include:

Reviewer: Since we are comparing this to bucket replication, the below almost implies that bucket replication doesn't do any of these.

I think if we want to compare-contrast this, it's closer to:

- mc batch replicate vs mc mirror - the advantages below
- mc batch replicate vs bucket replication - one time curated vs constant automatic

Author: Removed the direct comparison and altered the list of advantages some to.

.. _minio-batch-framework:

===============
Batch Framework
===============

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 2

.. versionchanged:: MinIO RELEASE.2022-10-08T20-11-00Z or later

>    The Batch Framework was introduced with the ``replicate`` job type in the :mc:`mc` RELEASES.2022-10-08T20-11-00Z.

Reviewer: You can probably just have `.. versionadded::` up top and leave it at that.

Author: Wait, you're telling me we have `.. versionchanged::` AND `.. versionadded::`?

Fixed.

Overview
--------

> The MinIO Batch Framework allows you to create, manage, monitor, and execute jobs using a YAML-formatted job definition file.

Reviewer: ```suggestion
The MinIO Batch Framework allows you to create, manage, monitor, and execute jobs using a YAML-formatted job definition file ("batch file").
```

Author: 👍

> The batch jobs run directly on the MinIO deployment to take advantage of the deployment's server-side processing power without constraints of the local machine where you run the :ref:`MinIO Client <minio-client>`.

Reviewer: ```suggestion
The batch jobs run directly on the MinIO deployment to take advantage of server-side processing power without constraints of the local machine where you run the :ref:`MinIO Client <minio-client>`.
```

Author: 👍

A batch file can define any number of job tasks.
MinIO does not limit on the number of job tasks that you can define in a batch file.

> Once started, MinIO begins processing the job immediately, depending on resources available to the deployment.

Reviewer: "immediately...depending" feels a little conflicting.

I am guessing it starts any and all jobs specified, and the completion time of those jobs depends on the available resources?

Author: Yeah, cleaned up.

> The job runs to completion.

Reviewer: I think we can cut this, this feels like it's stating the obvious.

Author: Agreed. Cut.

> If any portion of the job fails, MinIO retries the job up to the number of times defined in the job definition.

Reviewer: To confirm - if one job task fails, the remaining jobs continue processing.

Author: So each job in the batch becomes its own task with its own job ID. Each job ID runs to completion.

> A batch file only runs to completion one time.

Reviewer: MinIO processes each batch file *once* per execution. You can...

---
Something about this reads funny but I'm not sure what.

Author: I'm just cutting it. I think I was trying to explain something that was my own hangup/processing and not really anything that needs to be part of the doc.

You can :mc-cmd:`~mc batch start` the batch file as many times as desired to complete it again.

The MinIO Batch Framework supports the following job types:

.. list-table:: 
   :header-rows: 1
   :widths: 30 70
   :width: 100%   

   * - Job Type
     - Description

   * - ``replicate``
     - Perform a one-time replication procedure from one MinIO location to another MinIO location.

Prerequisites

> ~~~~~~~~~~~~~

Reviewer: Should be H1

Reviewer: Actually re-reading this section this is less of a `Prerequisites` and more of a `MinIO Batch CLI` or something.

Author: Renamed it to MinIO Batch CLI.

- Install the :ref:`MinIO Client <minio-client>`
- Define an :mc-cmd:`alias <mc alias set>` for the MinIO deployment

The :mc:`mc batch` commands include

.. list-table::
   :widths: 30 70
   :width: 90%

   * - :mc:`mc batch generate`
     - .. include:: /reference/minio-mc/mc-batch-generate.rst
          :start-after: start-mc-batch-generate-desc
          :end-before: end-mc-batch-generate-desc
   * - :mc:`mc batch start`
     - .. include:: /reference/minio-mc/mc-batch-start.rst
          :start-after: start-mc-batch-start-desc
          :end-before: end-mc-batch-start-desc
   * - :mc:`mc batch list`
     - .. include:: /reference/minio-mc/mc-batch-list.rst
          :start-after: start-mc-batch-list-desc
          :end-before: end-mc-batch-list-desc
   * - :mc:`mc batch status`
     - .. include:: /reference/minio-mc/mc-batch-status.rst
          :start-after: start-mc-batch-status-desc
          :end-before: end-mc-batch-status-desc
   * - :mc:`mc batch describe`
     - .. include:: /reference/minio-mc/mc-batch-describe.rst
          :start-after: start-mc-batch-describe-desc
          :end-before: end-mc-batch-describe-desc

Job Types
---------

``replicate``
~~~~~~~~~~~~~

> The ``replicate`` job type performs a single-run replication of objects from one MinIO deployment or location to another MinIO deployment or location.

Reviewer: 'deployment or location' ? latter might imply a non-MinIO path. I assume it's just MinIO -> MinIO and we can leave it at that.

Maybe an example of "For example, you can use a batch job to perform a one-time replication sync of objects from `minio-alpha/invoices/` to `minio-baker/invoices`"

Author: Added the example and removed location. Because it is MiniO to MinIO.

The definition file can limit the replication by bucket, prefix, and/or filters to only replicate certain objects.

> Key advantages of doing replication with a batch job include:

Reviewer: The advantages of Batch Replication over `mc mirror` include...

Author: 👍

- Removes the client to cluster network as a potential bottleneck

> - No special permissions for the user launching the job, as it runs entirely server side on the cluster

Reviewer: :thinking: do we not require any permissions for the Alias with which the command runs?

Author: Clarified with Harsha.
Added three new admin policies to the policy doc, added a security section to the framework doc, and reworked this line.

          :end-before: end-mc-batch-describe-desc

Access to ``mc batch``
----------------------

You can use MinIO's :ref:`Policy Based Access Control <minio-policy>` and the :ref:`administrative policy actions <minio-policy-mc-admin-actions>` to restrict who can start a batch job, retrieve a list of running jobs, or describe a running job.

Job Types
---------

``replicate``
~~~~~~~~~~~~~

The ``replicate`` job type performs a single-run replication of objects from one MinIO deployment to another.
The definition file can limit the replication by bucket, prefix, and/or filters to only replicate certain objects.

For example, you can use a batch job to perform a one-time replication sync of objects from ``minio-alpha/invoices/`` to ``minio-baker/invoices``.

The advantages of Batch Replication over :mc:`mc mirror` include:

- Removes the client to cluster network as a potential bottleneck
- A user only needs access to starting a batch job with no other permissions, as the job runs entirely server side on the cluster
- The job provides for retry attempts in event that objects do not replicate
- Batch jobs are one-time, curated processes allowing for fine control replication

> Sample YAML Description File for a ``replicate`` Job Type

Reviewer: Nit: I would just say "The following example batch template supports ``replicate`` job types"

Reviewer: with no header decorator. Or make it an H3 so it stubs in under ``replicate``

Author: Going with option 2 to give good room for the future when there are other job types.
